examples/flow_tool/schema.h

```diff
@@ -0,0 +1,112 @@
+
+namespace flow_tool {
+
+// Storage
+// =======
+
+namespace db {
+
+// `Blob` stores the contents single "file". It's versioned.
+CURRENT_ENUM(blob_key_t, uint64_t){InvalidBlob = 0ull};
+CURRENT_STRUCT(Blob) {
+  CURRENT_FIELD(key, blob_key_t, blob_key_t::InvalidBlob);  // The key of this blob.
+  CURRENT_FIELD(body, std::string);                         // The value of this blob.
+};
+
+// `Node` is an element of the hierarchical file system. Each node entry is a directory or a file.
+CURRENT_ENUM(node_key_t, uint64_t){InvalidNode = 0ull, RootNode = static_cast<uint64_t>(-1)};
+CURRENT_STRUCT(Dir) {
+  CURRENT_FIELD(dir, std::vector<node_key_t>);  // The list of files in this directory.
+};
+CURRENT_STRUCT(File) {
+  CURRENT_FIELD(blob, blob_key_t, blob_key_t::InvalidBlob);  // The reference to the respective blob.
+};
+CURRENT_STRUCT(Node) {
+  CURRENT_FIELD(key, node_key_t, node_key_t::InvalidNode);  // The unique key of this node.
+  CURRENT_FIELD(name, std::string);                         // The human-readable name of this file or dir, "" for root.
+  CURRENT_FIELD(data, (Variant<Dir, File>));                // The contents of this node.
+};
+
+// The storage for the above is maintained by class `FlowTool`.
+// It maintains some invariants, such as that the root is a directory and it can not be deleted.
+CURRENT_STORAGE_FIELD_ENTRY(UnorderedDictionary, Blob, PersistedBlob);
+CURRENT_STORAGE_FIELD_ENTRY(UnorderedDictionary, Node, PersistedNode);
+
+CURRENT_STORAGE(FlowToolDB) {
+  CURRENT_STORAGE_FIELD(blob, PersistedBlob);
+  CURRENT_STORAGE_FIELD(node, PersistedNode);
+};
+
+}  // namespace flow_tool::db
+
+// HTTP API
+// ========
+
+namespace api {
+
+CURRENT_STRUCT(Healthz) {
+  CURRENT_FIELD(up, bool, true);
+  CURRENT_FIELD(uptime, std::string);
+  CURRENT_FIELD(start_time_epoch_us, std::chrono::microseconds);
+  CURRENT_FIELD(init_time_us, std::chrono::microseconds);
+  CURRENT_FIELD(uptime_us, std::chrono::microseconds);
+};
+
+CURRENT_STRUCT(Error) {
+  CURRENT_FIELD(code, std::string);               // Some "CamelCase" error code.
+  CURRENT_FIELD(message, Optional<std::string>);  // Optional human-readable description.
+  CURRENT_CONSTRUCTOR(Error)
+  (std::string code = "UnknownError") : code(std::move(code)) {}
+  CURRENT_CONSTRUCTOR(Error)
+  (std::string code, std::string message) : code(std::move(code)), message(std::move(message)) {}
+};
+
+CURRENT_STRUCT(ErrorPathNotFound, Error) {
+  CURRENT_FIELD(error_component, std::string);
+  CURRENT_FIELD(full_path, std::vector<std::string>);
+  CURRENT_FIELD(index_in_full_path, uint32_t);
+};
+
+CURRENT_STRUCT(FileOrDirResponse) {
+  CURRENT_FIELD(url, std::string);
+  CURRENT_FIELD(path, std::vector<std::string>);
+};
+
+CURRENT_STRUCT(FileResponse, FileOrDirResponse) { CURRENT_FIELD(data, std::string); };
+
+CURRENT_STRUCT(DirResponse, FileOrDirResponse) { CURRENT_FIELD(dir, std::vector<std::string>); };
+
+}  // namespace flow_tool::api
+
+}  // namespace flow_tool
```

Comment:
Why is it called flow tool? It looks like a file system.

A short readme with motivation would make it clear.

Reply:
Because the tool is more than the file system. Readmes usually appear later, if/as the code becomes somehow useful :)

Reply:
Have you heard of README-driven development?

Reply:
No. Is it relevant at the moment?